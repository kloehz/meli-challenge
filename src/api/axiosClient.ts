import axios, { AxiosInstance } from "axios";

export const axiosClient: AxiosInstance = axios.create({
    baseURL: 'http://localhost:9000/api'
});