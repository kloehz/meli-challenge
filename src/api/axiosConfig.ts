import axios from "axios";

export const axiosConfig = axios.create({
    baseURL: 'http://localhost:9000/api'
});