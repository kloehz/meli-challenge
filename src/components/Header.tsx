import { memo, useState } from 'react';
import SearchImage from "../assets/images/SearchImage.svg";

interface ISetProducts {
  setProducts: React.Dispatch<React.SetStateAction<any[]>>
}

export const Header = memo(({setProducts}: ISetProducts) => {

  const [input, setInput] = useState('');

  const getProducts = () => {
    setInput(input.trim());

    if(input.length >= 1){
        console.log('resultado: ');
        setProducts(['product']);
    }
  }
  
  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    if(e.target.value[0] !== ' '){
      setInput(e.target.value);
    }
  }

  return (
    <div className="header">
      <img src="https://http2.mlstatic.com/frontend-assets/ui-navigation/5.16.1/mercadolibre/logo__large_plus.png" />
      <div className="input-container">
        <input
          type="text"
          className="input-header"
          placeholder="Nunca dejes de buscar"
          onChange={e => handleInputChange(e)}
          value={input}
        />
        <img
          src={SearchImage}
          className="search-icon"
          onClick={getProducts}
        />
      </div>
    </div>
    )
})
