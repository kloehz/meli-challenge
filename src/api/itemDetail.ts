import { IGetItemDetail } from "../types/getItemDetail";
import { axiosConfig } from "./axiosConfig";

export const apiGetItemDetails = (id: string) => {
  return axiosConfig.get<IGetItemDetail>(`/items/${id}`);
}
