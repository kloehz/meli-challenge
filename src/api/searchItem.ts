import { IGetItems } from "../types/getItems";
import { axiosConfig } from "./axiosConfig";

export const apiGetItems = (query: string) => {
  return axiosConfig.get<IGetItems>(`/items?q=:${query}`);
}
