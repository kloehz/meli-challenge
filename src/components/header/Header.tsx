import { useState, } from 'react';
import SearchImage from "../../assets/images/SearchImage.svg";
import { useHistory, useLocation } from "react-router-dom";
import { apiGetItems } from '../../api/searchItem';
import * as queryString from "query-string";


export const Header = () => {
  const [input, setInput] = useState('');
  const history = useHistory();
  const location  = useLocation()

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    setInput(input.trim());

    const { q: title } =  queryString.parse(location.search);
    const currentLocation = location.pathname.slice(1, location.pathname.length);
    const sameSearch = input === title && "items" === currentLocation;

    if(input.length >= 1 && !sameSearch){
      apiGetItems(input).then((res) => {
        history.push({pathname: "/items", search: `?q=${input}`, state: res.data })
      })
    }
  }
  
  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    if(e.target.value[0] !== ' '){
      setInput(e.target.value);
    }
  }

  const goHome = () => {
    setInput('');
    history.push('/');
  }

  return (
    <div className="header-container">
      <div className="header">
        <img
          src="https://http2.mlstatic.com/frontend-assets/ui-navigation/5.16.1/mercadolibre/logo__large_plus.png"
          onClick={goHome}
          className="logo-img"
        />
        <form className="input-container" onSubmit={handleSubmit}>
          <input
            type="text"
            className="input-header"
            placeholder="Nunca dejes de buscar"
            onChange={e => handleInputChange(e)}
            value={input}
          />
          <button type="submit" className="search-button">
            <img
              src={SearchImage}
              className="search-icon"
            />
          </button>
        </form>
      </div>
    </div>
    )
}
