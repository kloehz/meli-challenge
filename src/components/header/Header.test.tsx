import { shallow } from "enzyme"
import { apiGetItems } from "../../api/searchItem";
import { Header } from "./Header";

const mockHistoryPush = jest.fn()


jest.mock('react-router-dom', () => ({
    // @ts-ignore
    ...jest.requireActual("react-router-dom"),
    useLocation: jest.fn().mockReturnValue({
        pathname: '',
    }),
    useHistory: () => ({
      push: mockHistoryPush
    }),
}));

describe('Header Component tests', () => {
    
    const wrapper = shallow(<Header />);

    test('should be match snapshoot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should be NOT push to item page on click', () => {
        const fakeEvent = { preventDefault: () => {} };
        const formContainer = wrapper.find('.input-container');
        formContainer.simulate('submit', fakeEvent);
        expect(mockHistoryPush).not.toHaveBeenCalled();
    });

    test('should be change the value', () => {
        let input = wrapper.find('input');
        input.simulate('change', { target: {value: 'vasos'} });
        input = wrapper.find('input');
        expect(input.prop('value')).toBe('vasos');
    });

    test('should be display the correct image', () => {
        const img = wrapper.find('.logo-img');
        expect(img.prop('src')).toBe('https://http2.mlstatic.com/frontend-assets/ui-navigation/5.16.1/mercadolibre/logo__large_plus.png');
    });


    test('should be push to items page when submit', () => {
        const fakeEvent = { preventDefault: () => {} };
        let input = wrapper.find('input');
        input.simulate('change', { target: {value: 'vasos'} });
        input = wrapper.find('input');

        const formContainer = wrapper.find('.input-container');
        formContainer.simulate('submit', fakeEvent);
        expect(mockHistoryPush).not.toHaveBeenCalled();
    })
    
    
})
