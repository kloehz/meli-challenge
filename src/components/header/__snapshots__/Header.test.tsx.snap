// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component tests should be match snapshoot 1`] = `ShallowWrapper {}`;
